---
title: "Repositories"
author: "llrs"
date: "`r date()`"
output:
  html_document:
    toc: true
    toc_depth: 2
    toc_float: 
      collapsed: false
      smooth_scroll: false
---

<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-57101844-4', 'auto');
  ga('send', 'pageview');

</script>


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, collapse = TRUE)
```

# Projects

You can find [here](https://github.com/llrs?tab=repositories) my repositories. Some of them are projects worth mentioning. This projects are part of my daily work while others I enjoy developing at my own time. They are not ordered chronologically or by any special order. The titles are the link to the webpage or respository.

## [Bioconductor stats](https://llrs.github.com/BiocStats/) 

**Purpose**:  Analysing packages in Bioconductor by their downloads, IPs and the question for support.

This project is to compare how well the packages in Bioconductor work, track the raise and fall of package usage and how much does affect having a good support. 
    
## [WGCNA](https://github.com/llrs/WGCNA/) 

**Purpose**: Adding examples, tests and using roxygen in the WGCNA package.  

[WGCNA](https://labs.genetics.ucla.edu/horvath/CoexpressionNetwork/Rpackages/WGCNA/) let create weighted correlation network analysis. However it doesn't have test and the documentation I found it sometimes confusing (although abundant). With this project I learned how to contribute to a project (Avoid big modifications!!) and to read someone else code. 
    
##  [topGO](https://github.com/llrs/topGO) 

**Purpose**: Correcting some errors in the topGO package of Bioconductor.  

topGO is the only tool I found that takes into account the DAG structure of GO to calculate the enrichment scores. A fork of a prvious verion were I tryed to update and understand how it is implemented to correct bugs. The author's github page of the package is [here](https://github.com/adrianalexa/topGO). 
    
##  [BioCor](https://bioconductor.org/packages/BioCor) 

**Purpose**: A package to calculate functional similarities thorough pathways.  

I developed this package as part of my master thesis. You can find some how to use it in a [vignette](http://bioconductor.org/packages/release/bioc/vignettes/BioCor/inst/doc/BioCor.html) and the applications in [another one](http://bioconductor.org/packages/release/bioc/vignettes/BioCor/inst/doc/vignette2.html). Some other ideas to work with the package can be found [here](https://llrs.github.io/comments.html)


##  [GSEAdv](https://github.com/llrs/GSEAdv) 

**Purpose**: A package to analyse gene set collections.

It comes from the previous [comments](https://llrs.github.io/comments.html), and holds lots of code to analyze, simulate and compare different gene set collections, see also the [blog post](http://b101nfo.blogspot.com/2018/03/functional-enrichment-methods-and.html) about the history of the package. Using this software to comapre pathways databases will probably require another article.

##  [BaseSet](https://github.com/llrs/BaseSet) 

**Purpose**: A package to store set collections.

After finding some limitations on the GSEAdv project I started this project to test how sets could be stored efficiently. Later I found that a new system for holding gene sets was being considered. My idea is to end up merging GSEAdv methods into BaseSet. In this project I am using NSE (or tidy evaluations for the first time).
Some discussion and consideration about the new gene set class is beind written
[here](https://docs.google.com/document/d/1A3bs1rtbTo42Sgm9hPbLoG1lTGbQ-ITENaLRVyK2Njo/edit#)
    
##  [GOSemSim](https://github.com/GuangchuangYu/GOSemSim) 

**Purpose**: Calculate similarities between genes using [gene ontologies](http://www.geneontology.org/)

I contributed by speeding calculations in several functions, improving the implementation of the algorithms. I am officially a contributor of the [official repository](https://github.com/GuangchuangYu/GOSemSim).
    
##  [BrokerMaster](https://github.com/llrs/BrokerMaster) 

**Purpose**: A website to link projects and investigators with master students to improve the process in our master.

This was a project (for a subject in the master) to learn how to use databases and create websites.
    
## [Analysis of the thyroid carcinoma data from the TCGA project](https://llrs.github.com/IEO/IEO_report.html) 

**Purpose**: Find why women have more thyroid carcinoma.

This analysis was part of a subject for my master. 
We (it is a team work) explored several methods of batch correction and contrast. 
The respository with the data and output is [here](https://github.com/llrs/IEO).  
    
## [Rhea](https://github.com/llrs/Rhea) 

**Purpose**: Remove bugs for the [main repository](https://github.com/Lagkouvardos/Rhea).  

We are using this software as part of a collaboration, as it is aimed to be user-friendly I introduce  changes to help my team mates to use it. 

## [Thesis](https://llrs.github.com/thesis/) 

**Purpose**: Write my thesis along the way, instead of waiting for the last final months.

My thesis is about the integration of data, how the microbiome is related to the transcriptome in the inflammatory bowel disease.

## [experDesign](https://github.com/llrs/experDesign)

**Purpose**: A package for preparing the samples to be sequenced in batches

Divides the samples in equal batches, provides methods to select randomly but equilibrated samples. The name is from  expert design and experiment design.

## [RGCCA](https://github.com/llrs/RGCCA)

**Purpose**: Understand better the methods implemented and improve the efficiency.

A fork from CRAN was used to implement tests to the package, speed the functions and add some checks of the input parameters (Using this fork has already saved my some time).
I also added a function to check an input parameter.
